### PR TITLE
app: Split inbound/outbound constructors into components

### PR DESCRIPTION
**Pull request by olix0r**
This change does not change any functionality. It only restructures the inbound and outbound proxy modules so that the clients and servers can be instantiated separately. This will support gatewaying requests between the inbound and outbound proxy.

### CODE REVIEW

**Comment by olix0r** [comment]
Note that the majority of this change can be understood by reviewing `app/src/lib.rs`.  Also, fwiw, each commit in this PR should compile independently, in case this is easier to review chronologically.

File: linkerd/app/inbound/src/lib.rs [hunk 1]
```diff
@@ -52,252 +51,340 @@ pub struct Config {
...
+        );
+        self.build_server(
+            listen,
+            prevent_loop,
+            tcp_connect,
+            http_router,
+            local_identity,
+            metrics,
+            span_sink,
+            drain,
```
[the remark this answers is not in the document]
Comment (hawkw) [reply linkerd/app/inbound/src/lib.rs:87]: it's also a little weird to me that arguments we pass to both functions are passed in different orders, but again, probably nobody cares about this except me.

File: linkerd/app/inbound/src/lib.rs [hunk 2]
```diff
@@ -52,252 +51,340 @@ pub struct Config {
...
+        Error = impl Into<Error>,
+        Future = impl future::Future + Send,
+        Response = impl tokio::io::AsyncRead + tokio::io::AsyncWrite + Send + 'static,
+    > + tower::Service<
+        HttpEndpoint,
+        Error = impl Into<Error>,
+        Future = impl future::Future + Send,
+        Response = impl tokio::io::AsyncRead + tokio::io::AsyncWrite + Send + 'static,
+    > + Clone
+           + Send {
```
[the remark this answers is not in the document]
Comment (olix0r) [reply linkerd/app/inbound/src/lib.rs:105]: it is

File: linkerd/app/inbound/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (olix0r) [reply linkerd/app/inbound/src/lib.rs:87]: no, i think that's a valid point. I'll clean it up in the followup.

File: linkerd/app/outbound/src/lib.rs [hunk 3]
```diff
@@ -56,358 +56,417 @@ pub struct Config {
...
+            .into_inner()
+    }
+
+    pub fn build_dns_refine(
+        &self,
         dns_resolver: dns::Resolver,
+        metrics: &StackMetrics,
+    ) -> impl tower::Service<dns::Name, Response = dns::Name, Error = Error, Future = impl Send>
+           + Clone
+           + Send {
```
[the remark this answers is not in the document]
Comment (olix0r) [reply linkerd/app/outbound/src/lib.rs:93]: i don't know if rustfmt will allow that but i will try

[the remark this answers is not in the document]
Comment (olix0r) [reply linkerd/app/outbound/src/lib.rs:93]: it does not. i also agree all of this formatting is weird. but, take it up with the boss?

[the remark this answers is not in the document]
Comment (hawkw) [reply linkerd/app/outbound/src/lib.rs:93]: fair enough.

**Pull request merged**